app/models/concerns/pageflow/uploaded_file.rb

```diff
@@ -7,16 +7,45 @@ module UploadedFile
 
       belongs_to :uploader, :class_name => 'User'
       belongs_to :entry
+      belongs_to :parent_file, polymorphic: true
 
       has_many :usages, :as => :file, :class_name => 'Pageflow::FileUsage', :dependent => :destroy
       has_many :using_revisions, :through => :usages, :source => :revision
       has_many :using_entries, :through => :using_revisions, :source => :entry
       has_many :using_accounts, :through => :using_entries, :source => :account
+
+      validate :parent_allows_type_for_nesting
+    end
+
+    def parent_allows_type_for_nesting
+      if parent_file.present?
+        parent_class = parent_file_type.constantize
+        file_type_of_parent = Pageflow.config.file_types.find_by_model!(parent_class)
+        models_of_nested_file_types = file_type_of_parent.nested_file_types.map(&:model)
+        unless models_of_nested_file_types.include?(self.class)
+          errors[:base] << 'File type of provided parent file does not permit nesting files of '\
+                           "type #{self.class.name}"
+        end
+      end
     end
 
     # this prevents shadowing of usage_id that may be provided by SELECT alias
     def usage_id
       @usage_id || read_attribute(:usage_id)
     end
+
+    def nested_files(model)
+      model_table_name = model.table_name
+      model
+        .select("#{model_table_name}.*")
+        .where("#{model_table_name}.parent_file_id = #{id} AND "\
+              "#{model_table_name}.parent_file_type = '#{self.class.name}'")
```

Comment:
The column name `parent_file_type` leads to a confusing ambiguity, since Rails meands "(parent file) type", but it could also be read as "parent (file type)". In particular, there is no way of having a method now that would return the file type of the parent file, which leads to work arounds like parent_nested_file_types below.

[`belongs_to` supports a `:foreign_type` option](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-belongs_to) which we can use to give the column another name. I'd suggest using something like `parent_file_model_type`.